neo/SmartContract/Native/Tokens/NeoToken.cs

```diff
@@ -0,0 +1,422 @@
+using Neo.Cryptography.ECC;
+using Neo.IO;
+using Neo.Ledger;
+using Neo.Persistence;
+using Neo.VM;
+using Neo.VM.Types;
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Numerics;
+using VMArray = Neo.VM.Types.Array;
+
+namespace Neo.SmartContract.Native.Tokens
+{
+    public class NeoToken : NativeContractBase
+    {
+        public const string ServiceName = "Neo.Native.Tokens.NEO";
+        public static readonly byte[] Script = CreateNativeScript(ServiceName);
+        public static readonly UInt160 ScriptHash = Script.ToScriptHash();
+        public static readonly string[] SupportedStandards = { "NEP-5", "NEP-10" };
+        public const string Name = "NEO";
+        public const string Symbol = "neo";
+        public const int Decimals = 0;
+        public static readonly BigInteger DecimalsFactor = BigInteger.Pow(10, Decimals);
+        public static readonly BigInteger TotalAmount = 100000000 * DecimalsFactor;
+
+        private const byte Prefix_Initialized = 11;
+        private const byte Prefix_Account = 20;
+        private const byte Prefix_Validator = 33;
+        private const byte Prefix_ValidatorsCount = 15;
+
+        internal static bool Main(ApplicationEngine engine)
+        {
+            if (!new UInt160(engine.CurrentContext.ScriptHash).Equals(ScriptHash))
+                return false;
+            string operation = engine.CurrentContext.EvaluationStack.Pop().GetString();
+            VMArray args = (VMArray)engine.CurrentContext.EvaluationStack.Pop();
+            StackItem result;
+            switch (operation)
+            {
+                case "supportedStandards":
+                    result = SupportedStandards.Select(p => (StackItem)p).ToList();
+                    break;
+                case "name":
+                    result = Name;
+                    break;
+                case "symbol":
+                    result = Symbol;
+                    break;
+                case "decimals":
+                    result = Decimals;
+                    break;
+                case "totalSupply":
+                    result = TotalAmount;
+                    break;
+                case "balanceOf":
+                    result = BalanceOf(engine, args[0].GetByteArray());
+                    break;
+                case "transfer":
+                    result = Transfer(engine, args[0].GetByteArray(), args[1].GetByteArray(), args[2].GetBigInteger());
+                    break;
+                case "initialize":
+                    result = Initialize(engine);
+                    break;
+                case "unclaimedGas":
+                    result = UnclaimedGas(engine, args[0].GetByteArray(), (uint)args[1].GetBigInteger());
```

Comment:
We need the height parameter? maybe we should use our index

Reply:
That is true, I think that it is ok to use the current index of the CN's.

Reply:
We need deterministic results.

Reply:
But this subtract the gas that you receive previously, so deterministic is only possible while you sync the chain

Reply:
If you call `unclaimedGas` without `height`, when you received a block, for example block 1000, you don't know the result you get is for block `999` or `1000`. So we pass a height `999` for deterministic result.

Reply:
I think that among the consensus it is deterministic, @erikzhang.
If a node holds the transaction pending in the mempool that would be another thing.

Reply:
I think I didn't explain it clearly. When I call `unclaimedGas`, I get a result, but I don't know what the height this result corresponds to. Because the node may have just received a block when I called the method. In order to solve this problem, there are two ways. The first way is that I pass the `height` as a parameter to `unclaimedGas`. The second way is that `unclaimedGas` returns both the result and the corresponding height. I chose the first method because I can adjust the parameters.